File: packages/geometry/src/room.ts
/**
 * L1 room analysis (implementation-plan.md 5.2).
 *
 * The TypeScript twin of `workers/layout/roomfittr_layout/room.py`. It exists
 * because the editor validates every change in the browser (5.6, and 7.1
 * lists layout-edit validation as synchronous), and the answer it gives must
 * be the one the server will give when the edit is saved. If the two drift,
 * the editor lets a user build something the server then rejects.
 *
 * The circulation grid is the part that most easily drifts, so both sides
 * define it the same way: a cell is walkable when its centre is inside the
 * floor, at least a person's radius from any wall, and at least a person's
 * radius from every item. Those are distances, not buffered polygons --
 * a buffer is an approximation whose segment count is an implementation
 * detail, and two implementations would approximate it differently.
 */

import {
  type Footprint,
  type Point2,
  type Polygon,
  containsPoint,
  corners as cornersOf,
  distanceToBoundary,
  distanceToFootprint,
  rectangleFromWall,
} from "./geometry.ts";

/** 5.2: door keep-out is `width x width` into the room, plus a 900 mm approach. */
export const DOOR_APPROACH_MM = 900;
/** 5.2: the zone in front of a window that tall items should not occupy. */
export const WINDOW_ZONE_DEPTH_MM = 600;
/** 5.2 / H6: 50 mm cells, dilated by a 380 mm person radius (a 760 mm walkway). */
export const GRID_MM = 50;
export const PERSON_RADIUS_MM = 380;
/** H6: the share of free floor that must stay connected. */
export const MIN_CONNECTED_FRACTION = 0.6;
/** Nothing in the V1 vocabulary fits a shorter wall run than this. */
export const MIN_USABLE_RUN_MM = 400;

export interface Wall {
  readonly id: string;
  readonly start: Point2;
  readonly end: Point2;
  readonly kind: "solid" | "open";
}

export interface Opening {
  readonly id: string;
  readonly wallId: string;
  readonly type: "door" | "window" | "passage";
  readonly offsetMm: number;
  readonly widthMm: number;
  readonly sillMm: number;
  readonly heightMm: number;
  readonly swing: "left" | "right" | "unknown";
}

export interface Obstacle {
  readonly id: string;
  readonly label: string;
  readonly footprint: Footprint;
}

export interface FreeRun {
  readonly wallId: string;
  readonly startMm: number;
  readonly lengthMm: number;
}

export interface StaticGrid {
  readonly walkable: boolean[][];
  readonly gridX: number[][];
  readonly gridZ: number[][];
  readonly originXMm: number;
  readonly originZMm: number;
  readonly cellMm: number;
}

export interface CirculationGrid {
  readonly walkable: boolean[][];
  readonly originXMm: number;
  readonly originZMm: number;
  readonly cellMm: number;
}

export interface RoomAnalysis {
  readonly floor: Polygon;
  readonly walls: Wall[];
  readonly openings: Opening[];
  readonly obstacles: Obstacle[];
  readonly ceilingHeightMm: number;
  readonly doorKeepouts: Map<string, Footprint>;
  readonly windowZones: Map<string, Footprint>;
  readonly freeRuns: FreeRun[];
  readonly baseGrid: StaticGrid;
}

export function wallLength(wall: Wall): number {
  return Math.hypot(wall.end[0] - wall.start[0], wall.end[1] - wall.start[1]);
}

export function wallDirection(wall: Wall): Point2 {
  const length = wallLength(wall);
  if (length < 1e-9) return [0, 0];
  return [
    (wall.end[0] - wall.start[0]) / length,
    (wall.end[1] - wall.start[1]) / length,
  ];
}

/** Left normal, which points into the room for a CCW floor polygon. */
export function wallInwardNormal(wall: Wall): Point2 {
  const [dx, dz] = wallDirection(wall);
  return [-dz, dx];
}

export function wallPointAt(wall: Wall, offsetMm: number): Point2 {
  const [dx, dz] = wallDirection(wall);
  return [wall.start[0] + dx * offsetMm, wall.start[1] + dz * offsetMm];
}

/**
 * 5.2: a `width x width` swing area plus a 900 mm approach zone.
 *
 * When the swing is unknown -- which V1 almost always leaves it (3.6) -- the
 * keep-out covers a door's width either side rather than guessing a hinge.
 * Guessing wrong puts a bookcase where the door opens, and the user finds out
 * when the door hits it.
 */
export function doorKeepout(wall: Wall, opening: Opening): Footprint {
  const swingDepth = Math.max(opening.widthMm, DOOR_APPROACH_MM);
  if (opening.swing === "left" || opening.swing === "right") {
    return rectangleFromWall(
      wall.start,
      wall.end,
      opening.offsetMm,
      opening.widthMm,
      swingDepth,
    );
  }
  const start = Math.max(opening.offsetMm - opening.widthMm, 0);
  const end = Math.min(
    opening.offsetMm + 2 * opening.widthMm,
    wallLength(wall),
  );
  return rectangleFromWall(
    wall.start,
    wall.end,
    start,
    Math.max(end - start, 1),
    swingDepth,
  );
}

/** 5.2: the 600 mm-deep strip in front of a window. A warning zone, not a keep-out. */
export function windowZone(wall: Wall, opening: Opening): Footprint {
  return rectangleFromWall(
    wall.start,
    wall.end,
    opening.offsetMm,
    opening.widthMm,
    WINDOW_ZONE_DEPTH_MM,
  );
}

/**
 * The stretches of a wall with nothing in front of them.
 *
 * `open` walls return nothing: 3.6 says the solver treats them as a boundary
 * but never places against them, and a sofa backed onto thin air is exactly
 * the result that would embarrass the product.
 */
export function freeRuns(
  wall: Wall,
  openings: Opening[],
  obstacles: Obstacle[],
): FreeRun[] {
  const length = wallLength(wall);
  if (wall.kind !== "solid" || length < MIN_USABLE_RUN_MM) return [];

  const blocked: Array<[number, number]> = [];
  for (const opening of openings) {
    if (opening.wallId !== wall.id) continue;
    // A high window leaves the wall below it usable: a sideboard can sit
    // under it. Only floor-length openings interrupt the run.
    if (opening.type === "window" && opening.sillMm >= 700) continue;
    blocked.push([opening.offsetMm, opening.offsetMm + opening.widthMm]);
  }

  const [dx, dz] = wallDirection(wall);
  for (const obstacle of obstacles) {
    const ring = cornersOf(obstacle.footprint);
    const projections = ring.map(
      ([x, z]) => (x - wall.start[0]) * dx + (z - wall.start[1]) * dz,
    );
    const distances = ring.map(([x, z]) =>
      Math.abs((x - wall.start[0]) * -dz + (z - wall.start[1]) * dx),
    );
    if (Math.min(...distances) > 300) continue;
    const lo = Math.max(Math.min(...projections), 0);
    const hi = Math.min(Math.max(...projections), length);
    if (hi > lo) blocked.push([lo, hi]);
  }

  blocked.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
  const runs: FreeRun[] = [];
  let cursor = 0;
  for (const [lo, hi] of blocked) {
    if (lo - cursor >= MIN_USABLE_RUN_MM) {
      runs.push({ wallId: wall.id, startMm: cursor, lengthMm: lo - cursor });
    }
    cursor = Math.max(cursor, hi);
  }
  if (length - cursor >= MIN_USABLE_RUN_MM) {
    runs.push({ wallId: wall.id, startMm: cursor, lengthMm: length - cursor });
  }
  return runs;
}

/** Everything about the room's walkability that furniture cannot change. */
export function staticGrid(floor: Polygon, cellMm = GRID_MM): StaticGrid {
  let minX = Infinity;
  let minZ = Infinity;
  let maxX = -Infinity;
  let maxZ = -Infinity;
  for (const [x, z] of floor) {
    minX = Math.min(minX, x);
    minZ = Math.min(minZ, z);
    maxX = Math.max(maxX, x);
    maxZ = Math.max(maxZ, z);
  }
  const columns = Math.max(Math.ceil((maxX - minX) / cellMm), 1);
  const rows = Math.max(Math.ceil((maxZ - minZ) / cellMm), 1);

  const walkable: boolean[][] = [];
  const gridX: number[][] = [];
  const gridZ: number[][] = [];
  for (let row = 0; row < rows; row += 1) {
    const walkRow: boolean[] = [];
    const xRow: number[] = [];
    const zRow: number[] = [];
    const z = minZ + (row + 0.5) * cellMm;
    for (let column = 0; column < columns; column += 1) {
      const x = minX + (column + 0.5) * cellMm;
      xRow.push(x);
      zRow.push(z);
      const p: Point2 = [x, z];
      walkRow.push(
        containsPoint(floor, p) &&
          distanceToBoundary(p, floor) >= PERSON_RADIUS_MM,
      );
    }
    walkable.push(walkRow);
    gridX.push(xRow);
    gridZ.push(zRow);
  }
  return { walkable, gridX, gridZ, originXMm: minX, originZMm: minZ, cellMm };
}

/**
 * The walkable cells with `blocked` rectangles in place.
 *
 * A cell is blocked when its centre lies within a person's radius of any
 * item -- exactly the dilation a buffer would produce, computed as a distance
 * so this and the Python side mark the same cells.
 */
export function circulationGrid(
  blocked: Footprint[],
  base: StaticGrid,
): CirculationGrid {
  const walkable = base.walkable.map((row) => [...row]);
  for (const item of blocked) {
    if (item.widthMm <= 0 || item.depthMm <= 0) continue;
    for (let row = 0; row < walkable.length; row += 1) {
      const walkRow = walkable[row]!;
      const xRow = base.gridX[row]!;
      const zRow = base.gridZ[row]!;
      for (let column = 0; column < walkRow.length; column += 1) {
        if (!walkRow[column]) continue;
        if (
          distanceToFootprint([xRow[column]!, zRow[column]!], item) <
          PERSON_RADIUS_MM
        ) {
          walkRow[column] = false;
        }
      }
    }
  }
  return {
    walkable,
    originXMm: base.originXMm,
    originZMm: base.originZMm,
    cellMm: base.cellMm,
  };
}

export function cellOf(
  grid: CirculationGrid,
  xMm: number,
  zMm: number,
): [number, number] {
  return [
    Math.floor((zMm - grid.originZMm) / grid.cellMm),
    Math.floor((xMm - grid.originXMm) / grid.cellMm),
  ];
}

export function walkableCells(grid: CirculationGrid): number {
  let count = 0;
  for (const row of grid.walkable) for (const cell of row) if (cell) count += 1;
  return count;
}

/**
 * Flood fill from `seed` over walkable cells, 4-connected.
 *
 * 4-connected rather than 8-: diagonal moves would let a person slip between
 * two items touching at a corner, which they cannot do.
 */
export function connectedComponent(
  grid: CirculationGrid,
  seed: [number, number],
): boolean[][] {
  const rows = grid.walkable.length;
  const columns = rows > 0 ? grid.walkable[0]!.length : 0;
  const seen: boolean[][] = grid.walkable.map((row) => row.map(() => false));
  const [seedRow, seedColumn] = seed;
  if (seedRow < 0 || seedRow >= rows || seedColumn < 0 || seedColumn >= columns)
    return seen;
  if (!grid.walkable[seedRow]![seedColumn]) return seen;

  const queue: Array<[number, number]> = [seed];
  seen[seedRow]![seedColumn] = true;
  while (queue.length > 0) {
    const [row, column] = queue.shift()!;
    for (const [dr, dc] of [
      [1, 0],
      [-1, 0],
      [0, 1],
      [0, -1],
    ] as const) {
      const nr = row + dr;
      const nc = column + dc;
      if (nr < 0 || nr >= rows || nc < 0 || nc >= columns) continue;
      if (!grid.walkable[nr]![nc] || seen[nr]![nc]) continue;
      seen[nr]![nc] = true;
      queue.push([nr, nc]);
    }
  }
  return seen;
}

/** A point just inside the room in front of a door, where a person would stand. */
export function doorStandingPoint(wall: Wall, opening: Opening): Point2 {
  const centre = wallPointAt(wall, opening.offsetMm + opening.widthMm / 2);
  const [nx, nz] = wallInwardNormal(wall);
  const reach = PERSON_RADIUS_MM + GRID_MM;
  return [centre[0] + nx * reach, centre[1] + nz * reach];
}

/** The `RoomModel` document shape this module reads. */
export interface RoomModel {
  floor_polygon: Array<[number, number]>;
  ceiling_height_mm: number;
  walls: Array<{
    id: string;
    start: [number, number];
    end: [number, number];
    kind: "solid" | "open";
  }>;
  openings: Array<{
    id: string;
    wall_id: string;
    type: "door" | "window" | "passage";
    offset_mm: number;
    width_mm: number;
    sill_mm: number;
    height_mm: number;
    swing?: "left" | "right" | "unknown";
  }>;
  fixed_obstacles?: Array<{
    id: string;
    label: string;
    center: [number, number];
    size: [number, number, number];
    yaw_deg?: number;
  }>;
}

/** Run L1 over a `RoomModel` document. */
export function analyse(model: RoomModel): RoomAnalysis {
  const floor: Polygon = model.floor_polygon.map(([x, z]) => [x, z] as Point2);
  const walls: Wall[] = model.walls.map((raw) => ({
    id: raw.id,
    start: [raw.start[0], raw.start[1]] as Point2,
    end: [raw.end[0], raw.end[1]] as Point2,
    kind: raw.kind,
  }));
  const openings: Opening[] = model.openings.map((raw) => ({
    id: raw.id,
    wallId: raw.wall_id,
    type: raw.type,
    offsetMm: raw.offset_mm,
    widthMm: raw.width_mm,
    sillMm: raw.sill_mm,
    heightMm: raw.height_mm,
    swing: raw.swing ?? "unknown",
  }));
  const obstacles: Obstacle[] = (model.fixed_obstacles ?? []).map((raw) => ({
    id: raw.id,
    label: raw.label,
    footprint: {
      centerXMm: raw.center[0],
      centerZMm: raw.center[1],
      widthMm: raw.size[0],
      depthMm: raw.size[1],
      heightMm: raw.size[2],
      rotationDeg: raw.yaw_deg ?? 0,
      elevationMm: 0,
    },
  }));

  const byId = new Map(walls.map((wall) => [wall.id, wall]));
  const doorKeepouts = new Map<string, Footprint>();
  const windowZones = new Map<string, Footprint>();
  for (const opening of openings) {
    const wall = byId.get(opening.wallId);
    if (!wall) continue;
    if (opening.type === "door" || opening.type === "passage") {
      doorKeepouts.set(opening.id, doorKeepout(wall, opening));
    } else if (opening.type === "window") {
      windowZones.set(opening.id, windowZone(wall, opening));
    }
  }

  const runs: FreeRun[] = [];
  for (const wall of walls) runs.push(...freeRuns(wall, openings, obstacles));

  return {
    floor,
    walls,
    openings,
    obstacles,
    ceilingHeightMm: model.ceiling_height_mm,
    doorKeepouts,
    windowZones,
    freeRuns: runs,
    baseGrid: staticGrid(floor),
  };
}

export function wallById(
  analysis: RoomAnalysis,
  wallId: string,
): Wall | undefined {
  return analysis.walls.find((wall) => wall.id === wallId);
}

export function runsFor(analysis: RoomAnalysis, wallId: string): FreeRun[] {
  return analysis.freeRuns.filter((run) => run.wallId === wallId);
}

File: workers/layout/roomfittr_layout/room.py
"""L1 room analysis: what the room offers before anything is placed (5.2).

Everything here is derived deterministically from a `RoomModel`. The LLM never
sees this -- it gets the compact summary at the bottom -- and the solver and
the validator both read it, so a door keep-out means the same thing to the
thing that places furniture and the thing that judges it.

The circulation grid deserves a note. H6 requires that every door stays
reachable and that most of the floor stays connected, which is a question
about *paths*, not about distances between pairs of objects. Two items can
each be a comfortable distance from everything and still seal off a corner
between them. So the room is rasterised, the occupied cells are grown by a
person's radius, and connectivity is answered by flood fill -- which is the
only way to ask the question that is actually being asked.
"""

from __future__ import annotations

import math
from collections import deque
from dataclasses import dataclass, field
from typing import Any

import numpy as np
from numpy.typing import NDArray
from shapely.geometry import MultiPolygon, Polygon
from shapely.geometry.base import BaseGeometry
from shapely.ops import unary_union

from .geometry import Footprint, rectangle_from_wall

# 5.2: door keep-out is `width x width` into the room, plus a 900 mm approach.
DOOR_APPROACH_MM = 900.0

# 5.2: the zone in front of a window that tall items should not occupy.
WINDOW_ZONE_DEPTH_MM = 600.0

# 5.2 / H6: occupancy grid at 50 mm, dilated by a 380 mm person radius, which
# is the 760 mm minimum walkway H6 enforces.
GRID_MM = 50.0
PERSON_RADIUS_MM = 380.0

# H6: a layout must leave this share of the free floor connected. Below it,
# the room has been cut into pieces even if each piece is reachable from a door.
MIN_CONNECTED_FRACTION = 0.6

# A wall run shorter than this holds nothing in the V1 category vocabulary --
# the narrowest thing we place is a nightstand at ~400 mm.
MIN_USABLE_RUN_MM = 400.0


@dataclass(frozen=True, slots=True)
class Wall:
    id: str
    start: tuple[float, float]
    end: tuple[float, float]
    kind: str

    @property
    def length_mm(self) -> float:
        return math.dist(self.start, self.end)

    def direction(self) -> tuple[float, float]:
        length = self.length_mm
        if length < 1e-9:
            return (0.0, 0.0)
        return ((self.end[0] - self.start[0]) / length, (self.end[1] - self.start[1]) / length)

    def inward_normal(self) -> tuple[float, float]:
        """Left normal, which points into the room for a CCW floor polygon."""
        dx, dz = self.direction()
        return (-dz, dx)

    def point_at(self, offset_mm: float) -> tuple[float, float]:
        dx, dz = self.direction()
        return (self.start[0] + dx * offset_mm, self.start[1] + dz * offset_mm)


@dataclass(frozen=True, slots=True)
class Opening:
    id: str
    wall_id: str
    type: str
    offset_mm: float
    width_mm: float
    sill_mm: float
    height_mm: float
    swing: str = "unknown"


@dataclass(frozen=True, slots=True)
class Obstacle:
    id: str
    label: str
    footprint: Footprint


@dataclass(frozen=True, slots=True)
class FreeRun:
    """An uninterrupted stretch of a wall, available to place against."""

    wall_id: str
    start_mm: float
    length_mm: float

    @property
    def end_mm(self) -> float:
        return self.start_mm + self.length_mm


@dataclass(frozen=True, slots=True)
class CirculationGrid:
    """The room rasterised for connectivity questions."""

    walkable: NDArray[np.bool_]
    origin_x_mm: float
    origin_z_mm: float
    cell_mm: float

    def cell_of(self, x_mm: float, z_mm: float) -> tuple[int, int]:
        return (
            int((z_mm - self.origin_z_mm) // self.cell_mm),
            int((x_mm - self.origin_x_mm) // self.cell_mm),
        )

    @property
    def walkable_cells(self) -> int:
        return int(self.walkable.sum())


@dataclass(frozen=True, slots=True)
class RoomAnalysis:
    """Everything L1 derives. Read by the solver and by the validator."""

    floor: Polygon
    walls: list[Wall]
    openings: list[Opening]
    obstacles: list[Obstacle]
    ceiling_height_mm: float
    door_keepouts: dict[str, Polygon]
    window_zones: dict[str, Polygon]
    usable_floor: Polygon
    free_runs: list[FreeRun]
    warnings: list[str] = field(default_factory=list)

    @property
    def area_m2(self) -> float:
        return self.floor.area / 1e6

    def wall(self, wall_id: str) -> Wall | None:
        return next((w for w in self.walls if w.id == wall_id), None)

    def runs_for(self, wall_id: str) -> list[FreeRun]:
        return [run for run in self.free_runs if run.wall_id == wall_id]


def _footprint_from_obstacle(raw: dict[str, Any]) -> Footprint:
    size = raw["size"]
    return Footprint(
        center_x_mm=float(raw["center"][0]),
        center_z_mm=float(raw["center"][1]),
        width_mm=float(size[0]),
        depth_mm=float(size[1]),
        height_mm=float(size[2]),
        rotation_deg=float(raw.get("yaw_deg", 0.0)),
    )


def door_keepout(wall: Wall, opening: Opening) -> Polygon:
    """5.2: a `width x width` swing area plus a 900 mm approach zone.

    When the swing is unknown -- which V1 almost always leaves it (3.6) -- the
    keep-out covers the full width either side of the opening rather than
    guessing a hinge. Guessing wrong puts a bookcase where the door opens,
    and the user discovers it when the door hits it.
    """
    swing_depth = max(opening.width_mm, DOOR_APPROACH_MM)
    if opening.swing in ("left", "right"):
        return rectangle_from_wall(
            wall.start, wall.end, opening.offset_mm, opening.width_mm, swing_depth
        )

    # Unknown: widen along the wall by a door's width on each side, since the
    # leaf could sweep either way.
    start = max(opening.offset_mm - opening.width_mm, 0.0)
    end = min(opening.offset_mm + 2 * opening.width_mm, wall.length_mm)
    return rectangle_from_wall(wall.start, wall.end, start, max(end - start, 1.0), swing_depth)


def window_zone(wall: Wall, opening: Opening) -> Polygon:
    """5.2: the 600 mm-deep strip in front of a window.

    Not a keep-out. S3 only warns, and only for items tall enough to block
    the view -- a rug in front of a window is fine.
    """
    return rectangle_from_wall(
        wall.start, wall.end, opening.offset_mm, opening.width_mm, WINDOW_ZONE_DEPTH_MM
    )


def free_runs(wall: Wall, openings: list[Opening], obstacles: list[Obstacle]) -> list[FreeRun]:
    """The stretches of a wall with nothing in front of them.

    `open` walls return nothing: 3.6 says the solver treats them as a boundary
    but never places against them, and a sofa backed onto thin air is exactly
    the result that would embarrass the product.
    """
    if wall.kind != "solid" or wall.length_mm < MIN_USABLE_RUN_MM:
        return []

    blocked: list[tuple[float, float]] = []
    for opening in openings:
        if opening.wall_id != wall.id:
            continue
        if opening.type == "window" and opening.sill_mm >= 700.0:
            # A high window leaves the wall below it usable: a sideboard can
            # sit under it. Only floor-length openings interrupt the run.
            continue
        blocked.append((opening.offset_mm, opening.offset_mm + opening.width_mm))

    # An obstacle against this wall blocks the run behind it.
    dx, dz = wall.direction()
    for obstacle in obstacles:
        projections = [
            (x - wall.start[0]) * dx + (z - wall.start[1]) * dz
            for x, z in obstacle.footprint.corners()
        ]
        distances = [
            abs((x - wall.start[0]) * -dz + (z - wall.start[1]) * dx)
            for x, z in obstacle.footprint.corners()
        ]
        if min(distances) > 300.0:  # not against this wall
            continue
        lo, hi = max(min(projections), 0.0), min(max(projections), wall.length_mm)
        if hi > lo:
            blocked.append((lo, hi))

    blocked.sort()
    runs: list[FreeRun] = []
    cursor = 0.0
    for lo, hi in blocked:
        if lo - cursor >= MIN_USABLE_RUN_MM:
            runs.append(FreeRun(wall.id, cursor, lo - cursor))
        cursor = max(cursor, hi)
    if wall.length_mm - cursor >= MIN_USABLE_RUN_MM:
        runs.append(FreeRun(wall.id, cursor, wall.length_mm - cursor))
    return runs


def circulation_grid(
    floor: Polygon, blocked: list[Polygon], *, cell_mm: float = GRID_MM
) -> CirculationGrid:
    """Rasterise the room and grow the obstacles by a person's radius.

    Dilating the obstacles rather than shrinking the person is the standard
    configuration-space trick: it turns "can a 760 mm-wide person get
    through?" into "is there a path of free cells?", which a flood fill
    answers exactly.
    """
    min_x, min_z, max_x, max_z = floor.bounds
    columns = max(int(math.ceil((max_x - min_x) / cell_mm)), 1)
    rows = max(int(math.ceil((max_z - min_z) / cell_mm)), 1)

    xs = min_x + (np.arange(columns) + 0.5) * cell_mm
    zs = min_z + (np.arange(rows) + 0.5) * cell_mm
    grid_x, grid_z = np.meshgrid(xs, zs)

    # `Polygon.contains` per point is far too slow at this resolution --
    # a 5 x 4 m room is ~8000 cells. `contains_xy` is the vectorised form.
    from shapely import contains_xy

    inside: NDArray[np.bool_] = np.asarray(contains_xy(floor, grid_x, grid_z), dtype=bool)

    occupied: NDArray[np.bool_] = np.zeros((rows, columns), dtype=bool)
    if blocked:
        merged = unary_union([shape for shape in blocked if not shape.is_empty])
        if not merged.is_empty:
            # Dilate by the person radius, then mark the cells it covers.
            grown = merged.buffer(PERSON_RADIUS_MM)
            occupied = np.asarray(contains_xy(grown, grid_x, grid_z), dtype=bool)

    # The walls themselves are an obstacle: a person cannot stand with their
    # centre closer than their radius to a wall.
    near_wall: NDArray[np.bool_] = ~np.asarray(
        contains_xy(floor.buffer(-PERSON_RADIUS_MM), grid_x, grid_z), dtype=bool
    )

    walkable: NDArray[np.bool_] = inside & ~occupied & ~near_wall
    return CirculationGrid(walkable=walkable, origin_x_mm=min_x, origin_z_mm=min_z, cell_mm=cell_mm)


def connected_component(grid: CirculationGrid, seed: tuple[int, int]) -> NDArray[np.bool_]:
    """Flood fill from `seed` over walkable cells, 4-connected.

    4-connected rather than 8-: diagonal moves would let a person slip
    between two items touching at a corner, which they cannot do.
    """
    rows, columns = grid.walkable.shape
    seen = np.zeros_like(grid.walkable)
    if not (0 <= seed[0] < rows and 0 <= seed[1] < columns) or not grid.walkable[seed]:
        return seen

    queue = deque([seed])
    seen[seed] = True
    while queue:
        row, column = queue.popleft()
        for dr, dc in ((1, 0), (-1, 0), (0, 1), (0, -1)):
            nr, nc = row + dr, column + dc
            if 0 <= nr < rows and 0 <= nc < columns and grid.walkable[nr, nc] and not seen[nr, nc]:
                seen[nr, nc] = True
                queue.append((nr, nc))
    return seen


def door_standing_point(wall: Wall, opening: Opening) -> tuple[float, float]:
    """A point just inside the room in front of a door.

    Where a person entering would stand. H6 asks whether they can get from
    here to everywhere else, so the point has to be far enough in to clear
    the wall dilation.
    """
    centre = wall.point_at(opening.offset_mm + opening.width_mm / 2.0)
    nx, nz = wall.inward_normal()
    reach = PERSON_RADIUS_MM + GRID_MM
    return (centre[0] + nx * reach, centre[1] + nz * reach)


def analyse(room_model: dict[str, Any]) -> RoomAnalysis:
    """Run L1 over a `RoomModel` document."""
    floor = Polygon([(float(x), float(z)) for x, z in room_model["floor_polygon"]])
    if not floor.is_valid:
        floor = floor.buffer(0)

    walls = [
        Wall(
            id=raw["id"],
            start=(float(raw["start"][0]), float(raw["start"][1])),
            end=(float(raw["end"][0]), float(raw["end"][1])),
            kind=raw["kind"],
        )
        for raw in room_model["walls"]
    ]
    openings = [
        Opening(
            id=raw["id"],
            wall_id=raw["wall_id"],
            type=raw["type"],
            offset_mm=float(raw["offset_mm"]),
            width_mm=float(raw["width_mm"]),
            sill_mm=float(raw["sill_mm"]),
            height_mm=float(raw["height_mm"]),
            swing=raw.get("swing", "unknown"),
        )
        for raw in room_model["openings"]
    ]
    obstacles = [
        Obstacle(id=raw["id"], label=raw["label"], footprint=_footprint_from_obstacle(raw))
        for raw in room_model.get("fixed_obstacles", [])
    ]

    by_id = {wall.id: wall for wall in walls}
    door_keepouts: dict[str, Polygon] = {}
    window_zones: dict[str, Polygon] = {}
    for opening in openings:
        wall = by_id.get(opening.wall_id)
        if wall is None:
            continue
        if opening.type in ("door", "passage"):
            door_keepouts[opening.id] = door_keepout(wall, opening)
        elif opening.type == "window":
            window_zones[opening.id] = window_zone(wall, opening)

    blocking = [o.footprint.polygon() for o in obstacles] + list(door_keepouts.values())
    remaining: BaseGeometry = floor
    for shape in blocking:
        remaining = remaining.difference(shape)
    # A door keep-out can cut a corner off entirely. The largest piece is the
    # room; the offcut is a sliver behind the door that nothing fits in.
    usable = (
        max(remaining.geoms, key=lambda g: g.area)
        if isinstance(remaining, MultiPolygon)
        else remaining
    )

    runs: list[FreeRun] = []
    for wall in walls:
        runs.extend(free_runs(wall, openings, obstacles))

    warnings: list[str] = []
    if not any(o.type == "door" for o in openings):
        warnings.append("no_door")
    if not runs:
        warnings.append("no_free_wall")

    return RoomAnalysis(
        floor=floor,
        walls=walls,
        openings=openings,
        obstacles=obstacles,
        ceiling_height_mm=float(room_model["ceiling_height_mm"]),
        door_keepouts=door_keepouts,
        window_zones=window_zones,
        usable_floor=usable if isinstance(usable, Polygon) else floor,
        free_runs=runs,
        warnings=warnings,
    )


def llm_summary(analysis: RoomAnalysis, *, room_type_guess: str | None = None) -> dict[str, Any]:
    """The compact JSON the planner sends to Claude (5.2).

    Deliberately not a geometry dump. The LLM's job is to choose categories
    and spatial *intents* referring to ids; it never emits coordinates
    (Decision 4), so it does not need vertices -- and feeding it thousands of
    numbers would cost tokens and invite it to reason about geometry it is
    not the authority on.
    """
    walls: list[dict[str, Any]] = []
    for wall in analysis.walls:
        runs = analysis.runs_for(wall.id)
        has = [f"{o.id}:{o.type}" for o in analysis.openings if o.wall_id == wall.id]
        walls.append(
            {
                "id": wall.id,
                "kind": wall.kind,
                "length_mm": int(round(wall.length_mm)),
                "free_runs_mm": [int(round(run.length_mm)) for run in runs],
                "has": has,
            }
        )

    longest = max((run for run in analysis.free_runs), key=lambda r: r.length_mm, default=None)
    focal: list[str] = []
    window_walls = {o.wall_id for o in analysis.openings if o.type == "window"}
    focal.extend(f"{wall_id} (window wall)" for wall_id in sorted(window_walls))
    if longest is not None:
        focal.append(f"{longest.wall_id} (longest free run {longest.length_mm / 1000:.1f}m)")

    return {
        "room_type_guess": room_type_guess,
        "area_m2": round(analysis.area_m2, 1),
        "ceiling_mm": int(round(analysis.ceiling_height_mm)),
        "shape": _shape_of(analysis.floor),
        "walls": walls,
        "openings": [
            {
                "id": o.id,
                "type": o.type,
                "wall": o.wall_id,
                "width_mm": int(round(o.width_mm)),
            }
            for o in analysis.openings
        ],
        "fixed": [{"id": o.id, "label": o.label} for o in analysis.obstacles],
        "focal_candidates": focal,
    }


def _shape_of(floor: Polygon) -> str:
    corners = len(floor.exterior.coords) - 1
    if corners <= 4:
        return "rectangular"
    if corners <= 6:
        return "L"
    return "irregular"
